src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -192,41 +58,181 @@
+
+(defn- fmt
+  "Format a value using the provided formatter or identity function."
+  [formatter value]
+  ((or formatter identity) (common/format-value value)))
+
+(defn- build-column-headers
+  "Build multi-level column headers."
+  [{:keys [pivot-cols pivot-measures column-titles row-totals?]} col-combos col-formatters]
+  (concat
+   (if (= 1 (count pivot-measures))
+     (mapv (fn [col-combo] (mapv fmt col-formatters col-combo)) col-combos)
+     (for [col-combo col-combos
+           measure-key pivot-measures]
+       (conj
+        (mapv fmt col-formatters col-combo)
+        (get column-titles measure-key))))
+   (repeat (count pivot-measures)
+           (concat
+            (when (and row-totals? (seq pivot-cols)) ["Row totals"])
+            (repeat (dec (count pivot-cols)) nil)
+            (when (and (seq pivot-cols) (> (count pivot-measures) 1)) [nil])))))
+
+(defn- build-headers
+  "Combine row keys with column headers."
+  [column-headers {:keys [pivot-cols pivot-rows column-titles]}]
+  (map (fn [h]
+         (if (and (seq pivot-cols) (not (seq pivot-rows)))
+           (concat (map #(get column-titles %) pivot-cols) h)
+           (concat (map #(get column-titles %) pivot-rows) h)))
+       (let [hs (filter seq column-headers)]
+         (when (seq hs)
+           (apply map vector hs)))))
+
+(defn- build-row
+  "Build a single row of the pivot table."
+  [row-combo col-combos pivot-measures data totals row-totals? ordered-formatters row-formatters]
+  (let [row-path row-combo]
+    (concat
+     (when-not (seq row-formatters) (repeat (count pivot-measures) nil))
+     (mapv fmt row-formatters row-combo)
+     (concat
+      (for [col-combo col-combos
+            measure-key pivot-measures]
+        (fmt (get ordered-formatters measure-key)
+             (get-in data (concat row-path col-combo [measure-key]))))
+      (when row-totals?
+        (for [measure-key pivot-measures]
+          (fmt (get ordered-formatters measure-key)
+               (get-in totals (concat row-path [measure-key])))))))))
+
+(defn- build-column-totals
+  "Build column totals for a section."
+  [section col-combos pivot-measures totals row-totals? ordered-formatters pivot-rows]
+  (concat
+   (cons (format "Totals for %s" (fmt (get ordered-formatters (first pivot-rows)) section))
+         (repeat (dec (count pivot-rows)) nil))
+   (for [col-combo col-combos
+         measure-key pivot-measures]
+     (fmt (get ordered-formatters measure-key)
+          (get-in totals (concat
+                          [:column-totals section]
+                          col-combo
+                          [measure-key]))))
+   (when row-totals?
+     (for [measure-key pivot-measures]
+       (fmt (get ordered-formatters measure-key)
+            (get-in totals [:section-totals section measure-key]))))))
+
+(defn- build-grand-totals
+  "Build grand totals row."
+  [{:keys [pivot-cols pivot-rows]} col-combos pivot-measures totals row-totals? ordered-formatters]
```

Comment:
some of these build fns could probably do with a slight refactor of the order/grouping of args. (eg. should all of them take the config map as 1st arg, then row/col combos, then whatever else is needed.)